libs/langchain/langchain/document_loaders/readthedocs.py

```diff
@@ -73,6 +78,94 @@ def load(self) -> List[Document]:
                 docs.append(Document(page_content=text, metadata=metadata))
         return docs
 
+    def _get_link_ratio(self, section):
+        links = section.find_all("a")
+        total_text = "".join(str(s) for s in section.stripped_strings)
+        if len(total_text) == 0:
+            return 0
+
+        link_text = "".join(
+            str(string.string.strip())
+            for link in links
+            for string in link.strings
+            if string
+        )
+        return len(link_text) / len(total_text)
+
+    def _get_clean_text(self, element):
+        """
+        Recursive text getter that excludes code and binary content and
+        preserves newline data from the html.
+        """
+        from bs4 import NavigableString
+        from bs4.element import Comment
+
+        elements_to_skip = [
+            "script",
+            "noscript",
+            "canvas",
+            "meta",
+            "svg",
+            "map",
+            "area",
+            "audio",
+            "source",
+            "track",
+            "video",
+            "embed",
+            "object",
+            "param",
+            "picture",
+            "iframe",
+            "frame",
+            "frameset",
+            "noframes",
+            "applet",
+            "form",
+            "button",
+            "select",
+            "base",
+            "style",
+            "img",
+        ]
+
+        newline_elements = [
+            "p",
+            "div",
+            "ul",
+            "ol",
+            "li",
+            "h1",
+            "h2",
+            "h3",
+            "h4",
+            "h5",
+            "h6",
+            "pre",
+            "table",
+            "tr",
+        ]
+
+        def process_element(el):
```

Comment:
nit: preference to avoid inner functions. could we take this out of _get_clean_text and pass newline_elements and elements_to_skip in as args?